5.x Update remaining var annotation to use assert()

markstory opened the pull request:
There are still a bunch of `@var` annotations left, but they are for types more complex than PHP's typesystem can currently handle or were inline hints for factory methods.

markstory committed:
5.x Update remaining var annotation to use assert()

There are still a bunch of `@var` annotations left, but they are for
types more complex than PHP's typesystem can currently handle or were
inline hints for factory methods.

othercorey commented on src/Cache/Engine/FileEngine.php:
```diff
@@ -248,8 +248,8 @@ public function clear(): bool
         $cleared = [];
-        /** @var \SplFileInfo $fileInfo */
         foreach ($contents as $fileInfo) {
+            assert($fileInfo instanceof SplFileInfo);
```
This is fine and won't cost much, but I wish we could just check the array type once.

ADmad replied:
Doing the assert for each entry seems quite wasteful (even during development). This is how psalm itself deals with typing when iterating a recursive directory iterator https://github.com/vimeo/psalm/blob/b424de92d21df58d1ec62264550e20b6e39a3c85/src/Psalm/Config/FileFilter.php#L184-L185.

othercorey commented on src/Core/functions.php:
```diff
@@ -206,11 +206,10 @@ function env(string $key, string|float|int|bool|null $default = null): string|fl
         if ($val == null && getenv($key) !== false) {
-            /** @var string|false $val */
-            $val = getenv($key);
+            $val = (string)getenv($key);
```
How are we not testing this somehow?

markstory replied:
We probably haven't run into any of the cases that would trigger the value to not be a string.

markstory committed:
Appease the linting tool gods.

Humbly accept more metadata on what is going on as you're narrowing
types a bit too much or not enough.

ADmad commented on src/Cache/Cache.php:
```diff
@@ -169,8 +169,9 @@ protected static function _buildEngine(string $name): void
 
-            /** @var \Cake\Cache\CacheEngine $fallbackEngine */
             $fallbackEngine = clone static::pool($config['fallback']);
+            assert($fallbackEngine instanceof CacheEngine);
```
Seems this is required because of the `setConfig()` calls. If we need that method to be always available then it makes more sense to just add `setConfig()/getConfig()` to the `\Cake\Cache\CacheEngineInterface`.

ADmad commented on src/Core/PluginCollection.php:
```diff
@@ -285,8 +285,8 @@ public function create(string $name, array $config = []): PluginInterface
 
-            /** @var \Cake\Core\PluginInterface $plugin */
             $plugin = new $name($config);
+            assert($plugin instanceof PluginInterface);
```
I don't like such asserts just to please the static analyzers. All we are doing here is potentially changing a `TypeError` to `AssertError` which from a user POV makes no practical difference.

markstory replied:
That's fair. I'll revert these kind of changes.

markstory committed:
Reduce assert() usage

We don't need to assert() types when we'll also check types on return
values. Coercing filesystem iterators helps reduce overhead of type
checks as well.

markstory committed:
Merge branch '5.x' into ref-var-annotations
